src/server.js

```diff
@@ -0,0 +1,245 @@
+import http from 'http'
+import compose from 'koa-compose'
+import Executor from '@hydre/graphql-batch-executor'
+
+import {
+  on,
+} from 'events'
+
+import {
+  pipeline,
+} from 'stream'
+
+const log = debug('gql-ws').extend('server')
+const DEFAULT_TIMEOUT = 30_000
+
+export default class {
+  #ws_options
+  #schema
+  #rootValue
+  #context
+  #web_server
+
+  #middleware = []
+  #context_is_static
+  #timeout
+
+  /**
+   * Initialize a graphql websocket server with a configuration object
+   * @param {Object} options - server options
+   * @param {Number} [options.timeout=30000] - in millisecond,
+   *  try to deconnect non reachable clients
+   * @param {Object} [options.ws_options] - see https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketserveroptions-callback
+   * @param {graphql.SchemaDefinitionNode} options.schema - a graphql schema
+   * @param {Object} options.rootValue - a graphql resolver object
+   * @param {(Object|Function)} [options.context={}] - the context object
+   * or as a function/async function to dynamically
+   * build it on every upgrade request
+   */
+  constructor({
+    ws_options = {
+      path             : '/',
+      perMessageDeflate: false,
+      maxPayload       : 500,
+    },
+    schema = (() => {
+      throw new Error('Missing or invalid schema')
+    })(),
+    rootValue = {},
+    context = {},
+    web_server = http.createServer(),
+    timeout = DEFAULT_TIMEOUT,
+  }) {
+    this.#ws_options = ws_options
+    this.#schema = schema
+    this.#rootValue = rootValue
+    this.#web_server = web_server
+    this.#context = context
+    this.#timeout = timeout
+
+    this.#context_is_static = typeof context === 'object'
+  }
+
+  /**
+   * Resolve the context to allow usage of dynamic context building
+   * @param {(Object | Function | AsyncFunction)} context
+   */
+  context(...parameters) {
+    if (this.#context_is_static) return this.#context
+    return this.#context(...parameters)
+  }
+
+  use(middleware) {
+    this.#middleware.push(middleware)
+  }
+
+  listen(...parameters) {
+    const middleware = compose(this.#middleware)
+    const http_server = this.#web_server.listen(...parameters)
+    // we prevent usage of those as it's up to the http server to decide
+    // @see https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketserveroptions-callback
+    const {
+      // throwing off both vars
+      // eslint-disable-next-line no-unused-vars
+      host,
+      // eslint-disable-next-line no-unused-vars
+      port,
+      ...options
+    } = this.#ws_options
+    const wss = new WebSocket.Server({
+      ...options,
+      noServer: true,
+    })
+
+    wss.on('connection', (
+        ws, {
+          headers,
+        }, contextValue,
+    ) => {
+      const client_id = headers['sec-websocket-key']
+      const log_peer = log.extend(client_id)
+
+      log_peer('connected!')
+
+      const executor = new Executor({
+        id             : client_id,
```

Comment:
should add a comment here to inform that the id is just debug stuff and actually even remove it as it already is scoped under a ws

Reply:
still usefull to leg per client execution, may not impact anything actually